### PR TITLE
Add application-level per-realm firewall configuration

### PR DESCRIPTION
**Pull request by sethvargo**
This adds per-realm firewall settings (by allowed CIDR blocks) for the adminapi, apiserver, and main server. 

**Release Note**

```release-note
Add per-realm firewall security settings
```

/assign @mikehelmick

### CODE REVIEW

**Comment by mikehelmick** [comment]
/hold

I think this is dangerous for api server - that’s intended to be accessed by mobile devices

**Comment by sethvargo** [comment]
I wrote the help text to strongly discourage people from enabling the firewall against the apiserver, but I do think there's a legit use case where someone running their own server might want to do so. For example, they might be running a staging server and want to restrict devices on a particular subnet from uploading to it. One could argue they should add that protection at another layer, but I do see the use case.

**Comment by sherifkozman** [comment]
@sethvargo i believe it had to be on a different layer as well , where logging will be normally available in case the wrong CIDR block is added there will be no way to trace or debug.

File: cmd/server/assets/realmadmin/_form_security.html [hunk 1]
```diff
@@ -58,12 +61,60 @@
...
+      a certificate. It is highly recommended that you <strong>leave this
+      service publicly accessible</strong>. If blank, all traffic is allowed
+      from all IPs. These should be of <a
+      href="https://en.wikipedia.org/wiki/Classless_Inter-Domain_Routing"
+      target="_BLANK">CIDR notation</a>
+      of the format (e.g. <code>192.1.2.0/24</code>).
+    </small>
+  </div>
+
+  <div class="form-label-group">
```
Comment (mikehelmick) [inline cmd/server/assets/realmadmin/_form_security.html:103]: add a note that this is enforced POST login only?

File: pkg/controller/middleware/firewall.go [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+			}
+
+			var allowedCIDRs []string
+			switch typ {
+			case "adminapi":
+				allowedCIDRs = realm.AllowedCIDRsAdminAPI
+			case "apiserver":
+				allowedCIDRs = realm.AllowedCIDRsAPIServer
+			case "server":
+				allowedCIDRs = realm.AllowedCIDRsServer
```
Comment (mikehelmick) [inline pkg/controller/middleware/firewall.go:55]: case default:
   log - you did a bad?

File: pkg/controller/middleware/firewall.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+			ipStr := r.RemoteAddr
+
+			// Check if x-forwarded-for exists, the load balancer sets this, and the
+			// first entry is the real client IP.
+			xff := r.Header.Get("x-forwarded-for")
+			if xff != "" {
+				ipStr = strings.Split(xff, ",")[0]
+			}
+
+			ip := net.ParseIP(ipStr)
```
Comment (mikehelmick) [inline pkg/controller/middleware/firewall.go:76]: do we need to nil check ip?

File: pkg/database/migrations.go [hunk 4]
```diff
@@ -1275,6 +1275,27 @@ func (db *Database) getMigrations(ctx context.Context) *gormigrate.Gormigrate {
 					}
 				}
 
+				return nil
+			},
+		},
+		{
+			ID: "00052-CreateRealmAllowedCIDRs",
+			Migrate: func(tx *gorm.DB) error {
+				return tx.AutoMigrate(&Realm{}).Error
```
Comment (mikehelmick) [inline pkg/database/migrations.go:1284]: u did an auto migrate?!?!?

File: pkg/database/realm.go [hunk 5]
```diff
@@ -816,3 +823,34 @@ func (r *Realm) RenderWelcomeMessage() string {
...
+			// rudimentary.
+			if !strings.Contains(v, "/") {
+				if strings.Contains(v, ":") {
+					v = fmt.Sprintf("%s/128", v)
+				} else {
+					v = fmt.Sprintf("%s/32", v)
+				}
+			}
+
+			cidrs = append(cidrs, v)
```
Comment (mikehelmick) [inline pkg/database/realm.go:850]: should we ParseCIDR and check for errors here?

File: cmd/server/assets/realmadmin/_form_security.html (hunk 1 above)
Comment (sethvargo) [reply cmd/server/assets/realmadmin/_form_security.html:105]: Done

File: pkg/controller/middleware/firewall.go (hunk 2 above)
Comment (sethvargo) [reply pkg/controller/middleware/firewall.go:55]: Done

File: pkg/controller/middleware/firewall.go (hunk 3 above)
Comment (sethvargo) [reply pkg/controller/middleware/firewall.go:79]: Done

File: pkg/database/migrations.go (hunk 4 above)
Comment (sethvargo) [reply pkg/database/migrations.go:1284]: Don't tell

File: pkg/database/realm.go (hunk 5 above)
Comment (sethvargo) [reply pkg/database/realm.go:856]: Done

**Comment by sethvargo** [comment]
@sherifkozman the logs will include a message if a request is blocked due to the IP. It won't tell you _which_ IP, but you can correlate it to timestamps.

This can't be higher in the stack because it's a realm-level configuration, and users can be a member of multiple realms. For example, Jacky might be able to access realm1 from anywhere, but can only access realm2 from a corporate IP subnet.

**Comment by sethvargo** [comment]
@mikehelmick PTAnotherL

**Comment by mikehelmick** [comment]
> @mikehelmick PTAnotherL

👁️ looking

**Review by mikehelmick — APPROVED** [review]

**Pull request merged**